File: src/leetcode/p0200_0299/p0206_reverse_linked_list/ReverseLinkedList.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package org.willxu.algorithm.service.listnode;

import common.provide.ListNode;

public interface ReverseLinkedList {
    /**
     * Given the head of a singly linked list, reverse the list, and
     * return the reversed list.
     *
     * Follow up:
     * A linked list can be reversed either iteratively or recursively.
     * Could you implement both?
     *
     * 1. Iteratively:
     * {@link org.willxu.algorithm.service.impl.listnode.ReverseLinkedListIterator}
     *
     * @param head 0 <= head.size <= 5000
     *             -5000 <= Node.val <= 5000
     * @return head of the reversed list
     */
    ListNode reverseList(ListNode head);
}

File: src/leetcode/p0200_0299/p0206_reverse_linked_list/reverse_linked_list.py
from abc import ABC, abstractmethod
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class ReverseLinkedList(ABC):
    """
    Given the head of a singly linked list, reverse the list, and return the reversed list.
    """
    @abstractmethod
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        Args:
            head:
                - The number of nodes in the list is the range [0, 5000].
                - -5000 <= Node.val <= 5000
        """
        pass

class ReverseLinkedListRecursive(ReverseLinkedList):
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        last = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return last
